File: VB/WindowsApplication167/Array2DWrapper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Text
Imports System.Collections.Generic
Imports System.Collections
Imports System.ComponentModel

Namespace Example
	Public Class Array2DWrapper(Of T)
		Implements IList, ITypedList
		Private Class WrapperPropertyDescriptor
			Inherits PropertyDescriptor
			Private index As Integer
			Private elementType As Type

			Public Sub New(ByVal name As String, ByVal index As Integer, ByVal elementType As Type)
				MyBase.New(name, Nothing)
				Me.index = index
				Me.elementType = elementType
			End Sub
			Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
				Return False
			End Function

			Public Overrides ReadOnly Property ComponentType() As Type
				Get
					Return GetType(RowWrapper)
				End Get
			End Property

			Public Overrides Function GetValue(ByVal component As Object) As Object
				If TypeOf component Is RowWrapper Then
					Return (CType(component, RowWrapper)).GetValue(Me.index)
				End If
				Return Nothing
			End Function

			Public Overrides ReadOnly Property IsReadOnly() As Boolean
				Get
					Return False
				End Get
			End Property

			Public Overrides ReadOnly Property PropertyType() As Type
				Get
					Return elementType
				End Get
			End Property

			Public Overrides Sub ResetValue(ByVal component As Object)
			End Sub

			Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
				If TypeOf component Is RowWrapper Then
					CType(component, RowWrapper).SetValue(Me.index, value)
				End If
			End Sub

			Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
				Return False
			End Function
		End Class

		Private Class RowWrapper
			Inherits CustomTypeDescriptor
			Private owner As Array2DWrapper(Of T)

			Public Sub New(ByVal owner As Array2DWrapper(Of T))
				Me.owner = owner
			End Sub
			Public Overrides Overloads Function GetProperties() As PropertyDescriptorCollection
				Return MyBase.GetProperties()
			End Function

			Public Function GetValue(ByVal index As Integer) As Object
				Return owner.array.GetValue(owner.list.IndexOf(Me), index)
			End Function
			Public Sub SetValue(ByVal index As Integer, ByVal value As Object)
				owner.array.SetValue(value, owner.list.IndexOf(Me), index)
			End Sub

			Public Overrides Overloads Function GetProperties(ByVal attributes() As Attribute) As PropertyDescriptorCollection
				Return owner.pdc
			End Function
		End Class

		Private array(,) As T
		Private list As List(Of RowWrapper)

		Public Sub New(ByVal array(,) As T)
			Me.array = array
			Dim count As Integer = array.GetLength(0)
			list = New List(Of RowWrapper)(count)
			For i As Integer = 0 To count - 1
				list.Add(New RowWrapper(Me))
			Next i
		End Sub

		#Region "IList Members"

		Private Function Add(ByVal value As Object) As Integer Implements IList.Add
			Throw New Exception("The method or operation is not implemented.")
		End Function

		Private Sub Clear() Implements IList.Clear
			Throw New Exception("The method or operation is not implemented.")
		End Sub

		Private Function Contains(ByVal value As Object) As Boolean Implements IList.Contains
			If TypeOf value Is RowWrapper Then
				Return list.Contains(CType(value, RowWrapper))
			End If
			Return False
		End Function

		Private Function IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
			If TypeOf value Is RowWrapper Then
				Return list.IndexOf(CType(value, RowWrapper))
			End If
			Return -1
		End Function

		Private Sub Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
			Throw New Exception("The method or operation is not implemented.")
		End Sub

		Private ReadOnly Property IsFixedSize() As Boolean Implements IList.IsFixedSize
			Get
				Return True
			End Get
		End Property

		Private ReadOnly Property IsReadOnly() As Boolean Implements IList.IsReadOnly
			Get
				Return True
			End Get
		End Property

		Private Sub Remove(ByVal value As Object) Implements IList.Remove
			Throw New Exception("The method or operation is not implemented.")
		End Sub

		Private Sub RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
			Throw New Exception("The method or operation is not implemented.")
		End Sub

		Public Property IList_Item(ByVal index As Integer) As Object Implements IList.Item
			Get
				Return list(index)
			End Get
			Set(ByVal value As Object)
				Throw New Exception("The method or operation is not implemented.")
			End Set
		End Property

		#End Region

		#Region "ICollection Members"

		Private Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
			If TypeOf array Is RowWrapper() Then
				list.CopyTo(CType(array, RowWrapper()), index)
			End If
		End Sub

		Private ReadOnly Property Count() As Integer Implements ICollection.Count
			Get
				Return list.Count
			End Get
		End Property

		Private ReadOnly Property IsSynchronized() As Boolean Implements ICollection.IsSynchronized
			Get
				Return False
			End Get
		End Property

		Private ReadOnly Property SyncRoot() As Object Implements ICollection.SyncRoot
			Get
				Return Me
			End Get
		End Property

		#End Region

		#Region "IEnumerable Members"

		Private Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
			Return list.GetEnumerator()
		End Function

		#End Region

		#Region "ITypedList Members"

		Private pdc As PropertyDescriptorCollection

		Private Function GetItemProperties(ByVal listAccessors() As PropertyDescriptor) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
			If pdc Is Nothing Then
				Dim pd(array.GetLength(1) - 1) As PropertyDescriptor
				For i As Integer = 0 To pd.Length - 1
					pd(i) = New WrapperPropertyDescriptor(AscW("C" )+ i, i, GetType(T))
				Next i
				pdc = New PropertyDescriptorCollection(pd)
			End If
			Return pdc
		End Function

		Private Function GetListName(ByVal listAccessors() As PropertyDescriptor) As String Implements ITypedList.GetListName
			Return String.Empty
		End Function

		#End Region

	End Class
End Namespace

File: CS/WindowsApplication167/Array2DWrapper.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Collections;
using System.ComponentModel;

namespace Example
{
    public class Array2DWrapper<T> : IList, ITypedList
    {
        class WrapperPropertyDescriptor : PropertyDescriptor
        {
            int index;
            Type elementType;

            public WrapperPropertyDescriptor(string name, int index, Type elementType)
                : base(name, null)
            {
                this.index = index;
                this.elementType = elementType;
            }
            public override bool CanResetValue(object component)
            {
                return false;
            }

            public override Type ComponentType
            {
                get { return typeof(RowWrapper); }
            }

            public override object GetValue(object component)
            {
                if (component is RowWrapper) return ((RowWrapper)component).GetValue(this.index);
                return null;
            }

            public override bool IsReadOnly
            {
                get { return false; }
            }

            public override Type PropertyType
            {
                get { return elementType; }
            }

            public override void ResetValue(object component)
            {
            }

            public override void SetValue(object component, object value)
            {
                if (component is RowWrapper) ((RowWrapper)component).SetValue(this.index, value);
            }

            public override bool ShouldSerializeValue(object component)
            {
                return false;
            }
        }

        class RowWrapper : CustomTypeDescriptor
        {
            Array2DWrapper<T> owner;

            public RowWrapper(Array2DWrapper<T> owner)
            {
                this.owner = owner;
            }
            public override PropertyDescriptorCollection GetProperties()
            {
                return base.GetProperties();
            }

            public object GetValue(int index)
            {
                return owner.array.GetValue(owner.list.IndexOf(this), index);
            }
            public void SetValue(int index, object value)
            {
                owner.array.SetValue(value, owner.list.IndexOf(this), index);
            }

            public override PropertyDescriptorCollection GetProperties(Attribute[] attributes)
            {
                return owner.pdc;
            }
        }

        private T[,] array;
        private List<RowWrapper> list;

        public Array2DWrapper(T[,] array)
        {
            this.array = array;
            int count = array.GetLength(0);
            list = new List<RowWrapper>(count);
            for (int i = 0; i < count; i++)
                list.Add(new RowWrapper(this));
        }

        #region IList Members

        int IList.Add(object value)
        {
            throw new Exception("The method or operation is not implemented.");
        }

        void IList.Clear()
        {
            throw new Exception("The method or operation is not implemented.");
        }

        bool IList.Contains(object value)
        {
            if (value is RowWrapper) return list.Contains((RowWrapper)value);
            return false;
        }

        int IList.IndexOf(object value)
        {
            if (value is RowWrapper) return list.IndexOf((RowWrapper)value);
            return -1;
        }

        void IList.Insert(int index, object value)
        {
            throw new Exception("The method or operation is not implemented.");
        }

        bool IList.IsFixedSize
        {
            get { return true; }
        }

        bool IList.IsReadOnly
        {
            get { return true; }
        }

        void IList.Remove(object value)
        {
            throw new Exception("The method or operation is not implemented.");
        }

        void IList.RemoveAt(int index)
        {
            throw new Exception("The method or operation is not implemented.");
        }

        object IList.this[int index]
        {
            get
            {
                return list[index];
            }
            set
            {
                throw new Exception("The method or operation is not implemented.");
            }
        }

        #endregion

        #region ICollection Members

        void ICollection.CopyTo(Array array, int index)
        {
            if (array is RowWrapper[]) list.CopyTo((RowWrapper[])array, index);
        }

        int ICollection.Count
        {
            get { return list.Count; }
        }

        bool ICollection.IsSynchronized
        {
            get { return false; }
        }

        object ICollection.SyncRoot
        {
            get { return this; }
        }

        #endregion

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator()
        {
            return list.GetEnumerator();
        }

        #endregion

        #region ITypedList Members

        PropertyDescriptorCollection pdc;

        PropertyDescriptorCollection ITypedList.GetItemProperties(PropertyDescriptor[] listAccessors)
        {
            if (pdc == null)
            {
                PropertyDescriptor[] pd = new PropertyDescriptor[array.GetLength(1)];
                for (int i = 0; i < pd.Length; i++)
                {
                    pd[i] = new WrapperPropertyDescriptor("C" + i, i, typeof(T));
                }
                pdc = new PropertyDescriptorCollection(pd);
            }
            return pdc;
        }

        string ITypedList.GetListName(PropertyDescriptor[] listAccessors)
        {
            return string.Empty;
        }

        #endregion

    }
}